### PR TITLE
Prefer depth-first plan evaluation for unions and joins

### PR DESCRIPTION
**Pull request by lnkuiper**
This PR adds a way to recursively introduce dependencies in `MetaPipeline` to avoid evaluating query plans breadth-first and instead prefer depth-first.

For now, I've added a simple heuristic. If the LHS child of a union can likely keep all threads busy, we let the RHS depend on it (recursively). This should fix issues where we have query plans with many unions that all use a lot of memory. By not fully evaluating the children, temporary query intermediates pile up, and more offloading is needed than necessary. I think this fixes one of your issues @Alex-Monahan @bleskes.

I've done the same thing for joins. If we have a wide, bushy query plan, we also prefer to evaluate the plan in a depth-first fashion, as this reduces memory pressure. I've used the same heuristic there.

I've added a test for the union and the join, which used to fail but now succeeds with the new dependencies.

### CODE REVIEW

**Comment by Alex-Monahan** [comment]
This is fantastic!! Thank you Laurens!

**Comment by lnkuiper** [comment]
Seems like I was a bit eager in sending the PR on a Friday afternoon. I'll fix it up on Monday!

**Comment by lnkuiper** [comment]
> Seems like I was a bit eager in sending the PR on a Friday afternoon. I'll fix it up on Monday!

Way too eager! I spent a ton of time fixing queries with a combination of unions and joins, fingers crossed for CI now. I've also added some code to make sure that blocks that will be destroyed are not added to the block eviction queue

File: src/common/types/row/tuple_data_allocator.cpp [hunk 1]
```diff
@@ -5,6 +5,8 @@
 #include "duckdb/common/types/row/tuple_data_states.hpp"
 #include "duckdb/storage/buffer_manager.hpp"
 
+#include <duckdb/storage/buffer/block_handle.hpp>
```
Comment (Maxxen) [inline src/common/types/row/tuple_data_allocator.cpp:8]: Looks like someone else also updated CLion ;)

Comment (lnkuiper) [reply src/common/types/row/tuple_data_allocator.cpp]: Ha, thanks for catching this. I hadn't noticed CLion was sabotaging me

**Comment by lnkuiper** [comment]
"Times two" really helped! Query 4 and 11 still regress slightly:
```
benchmark/tpcds/sf1/q04.benchmark
Old timing: 0.377278
New timing: 0.527084
...
benchmark/tpcds/sf1/q11.benchmark
Old timing: 0.239347
New timing: 0.3037
```
But many other queries improved slightly, so the overall geometric mean improved!
```
Old timing geometric mean: 0.040694049633221076
New timing geometric mean: 0.04044983206934163
```

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**